File: LeetCode/LeetCode 78. 子集 (dfs).py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        ans = list()
        ss = list()

        def dfs(x: int, n: int):
            if x == n:
                ans.append(ss[:])
                return

            ss.append(nums[x])
            dfs(x + 1, n)
            ss.pop()
            dfs(x + 1, n)

        dfs(0, n)
        return ans

File: LeetCode/LeetCode 78. 子集 (dfs).cpp
class Solution {
public:
    int n;
    vector <int> ss;
    vector <vector<int>> ans;

    void dfs(int x, vector<int>& nums) {
        if(x == n) {
            ans.push_back(ss);
            return ;
        }
        ss.push_back(nums[x]);
        dfs(x + 1, nums);
        ss.pop_back();
        dfs(x + 1, nums);
        return ;
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        n = nums.size();
        dfs(0, nums);
        return ans;
    }
};